### PR TITLE
Adding a broken attempt at injecting knex with types into a12n

### PR DESCRIPTION
**Pull request by mihok**
### Changes

Im attempting to include knex as the database client to enable PostgreSQL in addition to MySQL, and in the future enable all the database clients knex supports.

Currently trying to keep up with types, and providing a minimal attempt at not having to create any breaking changes.

I'm currently fighting with the `params` parameter for the `.query` function (which now just pipes directly into `knex.raw`. Their function signatures match, and in theory the types they use should match as well but I'm still having issues. (reference to all the knex types: https://github.com/knex/knex/blob/master/types/index.d.ts)

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -27,8 +26,6 @@ if (!process.env.PUBLIC_URI) {
 (async () => {
 
   console.log('Connecting to database');
-  await checkPatches();
```
Comment (evert) [inline src/app.ts:30]: I think we need to auto-run knex migrations on start. I believe knex ensures that if multiple a12nserver start at the same time, only 1 will run the migration (using table locks).

File: src/database.ts [hunk 2]
```diff
@@ -1,88 +1,77 @@
...
-  let settings: any = {};
-
-  // We are running in a local environment.
-  settings = {
-    user: process.env.MYSQL_USER,
-    password: process.env.MYSQL_PASSWORD,
-    database: process.env.MYSQL_DATABASE,
-  };
+  let connection: any = {};
+  let client: string = '';
```
Comment (evert) [inline src/database.ts:35]:
```suggestion
  let connection;
  let client;
```

I assume if you don't initialize these, typescript will complain. That's a good thing, because I think you just need an `else` statement with an explicit error.

File: src/database.ts [hunk 3]
```diff
@@ -1,88 +1,77 @@
...
 }
 
 export default {
-  query,
   getConnection,
-  checkPatches,
+  query,
 };
 
 async function getSettings() {
```
Comment (evert) [inline src/database.ts:32]: Could this return something like `Knex.Config` or `Knex.Settings` (I assume a type exists)

File: src/database.ts [hunk 4]
```diff
@@ -1,88 +1,77 @@
...
+      connection.socketPath = '/cloudsql/' + process.env.MYSQL_INSTANCE_CONNECTION_NAME;
+    } else {
+      delete connection.host;
+      delete connection.port;
+    }
   }
-  return settings;
 
-}
+  // We are running in a local environment.
```
Comment (evert) [inline src/database.ts:65]: I don't understand this comment =)

File: src/mfa/totp/service.ts [hunk 5]
```diff
@@ -17,17 +17,19 @@ export function generateSecret(): string {
...
     user_id: totpDevice.user.id,
     secret: totpDevice.secret,
   };
 
-  const result = await database.query(query, [newTotpDeviceRecord]);
+  const connection = await database.getConnection();
+
+  const result = await connection<UserTotpDeviceRow>('user_totp')
+    .insert(newTotpDeviceRecord, 'id')
+    .returning('id')
```
Comment (evert) [inline src/mfa/totp/service.ts:29]: I'm fairly certain `.returning()`  breaks in MySQL. I recently ran into this myself. Knex (for me) throws a warning that it's not supported in MySQL.

Comment (evert) [reply src/mfa/totp/service.ts:29]: Also this might be missing a value for the `created` field

File: src/mfa/webauthn/service.ts [hunk 6]
```diff
@@ -15,31 +15,34 @@ type UserWebAuthnRow = {
...
       public_key: webAuthNDevice.publicKey.toString('base64'),
       counter: webAuthNDevice.counter
     };
 
-    const result = await database.query(query, [newWebAuthnRecord]);
+    const connection = await database.getConnection();
+
+    const result = await connection<UserWebAuthnRow>('user_webauthn')
+      .insert(newWebAuthnRecord, 'id')
+      .returning('id');
```
Comment (evert) [inline src/mfa/webauthn/service.ts:29]: * [ ] created missing?
* [ ] returning might not work

File: src/oauth2-client/service.ts [hunk 7]
```diff
@@ -118,26 +118,27 @@ export async function getOAuth2ClientFromBody(ctx: Context<any>): Promise<OAuth2
...
+  const connection = await db.getConnection();
+  const result = await connection<OAuth2ClientRecord>('oauth2_clients')
+    .insert(params, 'id')
+    .returning('id');
+
+  const newClient = await connection<OAuth2ClientRecord, OAuth2ClientRecord>('oauth2_client')
+    .select({ id: result[0] })
+    .returning('*');
+
+  const realClient = await mapToOauth2Client(newClient[0], client.app, client.allowedGrantTypes);
```
Comment (evert) [inline src/oauth2-client/service.ts:137]: Doesn't need async/await

File: src/oauth2-client/service.ts [hunk 8]
```diff
@@ -152,3 +153,16 @@ export async function validateSecret(oauth2Client: OAuth2Client, secret: string)
...
 
+async function mapToOauth2Client(data: OAuth2ClientRecord, app: App, allowedGrantTypes: GrantType[]): Promise<OAuth2Client> {
+  return {
+    id: data.id,
+    clientId: data.client_id,
+    clientSecret: data.client_secret.toString(),
+    app: {
+      ...app,
+      id: data.user_id,
+    },
```
Comment (evert) [inline src/oauth2-client/service.ts:164]: Would this work?

```suggestion
    app,
```

File: src/database.ts (hunk 3 above)
Comment (evert) [reply src/database.ts:32]: I don't think this function needs to be `async`. Maybe this was wrong before.

File: src/app.ts (hunk 1 above)
Comment (evert) [reply src/app.ts:30]: I think we need to migrate the migrations to knex.

I don't think this means a full-rewrite of all migrations to knex migrations, but I do think we need a knex migration that:

1. Checks the patch level
2. If it's below the latest patch (from the old patch system), either:
  * complain and die and tell users to manually apply these patches
  * Or: automatically run the patches (with `.raw` ?)

This is to support an upgrade path from an earlier a12nserver.

**Review by evert — CHANGES_REQUESTED** [review]
What would be really nice is if:

* We could get some github actions test that sets up a Postgres or MySQL and see if it can start a12nserver, maybe a rudimentary test?
* Types for knex records, see: https://github.com/evert/tt/blob/main/src/knex.ts

File: src/mfa/totp/service.ts (hunk 5 above)
Comment (mihok) [reply src/mfa/totp/service.ts:29]: Yeah, so line 28's additional parameter is what allows MySQL to return as expected, so to support both databases you have to do line 28 and line 29, `.returning('id')` wont return the id in MySQL on its own.

File: src/database.ts (hunk 2 above)
Comment (evert) [reply src/database.ts:52]: This is not 100% resolved, because initializing these variables is still a bad idea.

File: src/app.ts (hunk 1 above)
Comment (mihok) [reply src/app.ts:30]: Hmmm, I think I might go with the complain and die approach for now...

Comment (evert) [reply src/app.ts:30]: Yeah i think that's fine. This will at least prevent people from starting the server with a broken database.

File: src/database.ts (hunk 2 above)
Comment (mihok) [reply src/database.ts:52]: Ohhh I didnt fully understand your initial comment before! I'll fix

File: Makefile [hunk 9]
```diff
@@ -20,15 +20,24 @@ test:
...
 lint-fix: fix
 
+migrate:
+	npx knex migrate:latest
+
+migrate-rollback:
+	npx knex migrate:rollback
+
+migration:
+	npx knex migrate:make $(name) -x ts
```
Comment (evert) [inline Makefile:37]: :+1:

File: Makefile [hunk 10]
```diff
@@ -43,5 +52,7 @@ dist/build: $(SOURCE_FILES)
 	touch dist/build
 
 inspect: build
-	node --inspect dist/app.js
+	node --inspect dist/src/app.js
```
Comment (evert) [inline Makefile:55]: I hope we can revert this change. I ran into the same issue and it was because `knexfile.ts` is in the root.

Can we move `knexfile.ts` into `src/` ? The `npx knex` command has a `--knexfile` option to point to the right place.

I also did this here:

https://github.com/badgateway/tt/blob/main/package.json#L14

File: knexfile.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+// Update with your config settings.
```
Comment (evert) [inline knexfile.ts:1]:
```suggestion
```

File: migrations/20220105195635_init.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    if (result[0].id < MAGIC_CHANGELOG_NUMBER) {
+      throw new Error('Migrations must have reached `changelog` version 42 before applying Knex migrations, please apply those manually and then run `make migrate`.');
+    }
+
+  } catch (err: any) {
+    // Color the error red.
+    // eslint-disable-next-line no-console
+    console.error('\x1b[31m', err, '\x1b[37m');
+
+    process.exit(1);
```
Comment (evert) [inline migrations/20220105195635_init.ts:27]: Instead of `exit,` you can also just `throw err` or just never catch this error at all (Knex should handle errors)

File: src/app.ts [hunk 13]
```diff
@@ -27,8 +26,7 @@ if (!process.env.PUBLIC_URI) {
...
-
+  console.log('Running Migrations');
```
Comment (evert) [inline src/app.ts:29]: Is this a lie?

**Review by evert — COMMENTED** [review]
Just a few more requests.

Great PR. It doesn't just add major features (postgres support, better migrations) but also improves most of the code it touches

File: src/app.ts (hunk 13 above)
Comment (mihok) [reply src/app.ts:29]: Lol it happens when the pool connection is made but seemed weird with the logs going:

1. Connecting to database
2. Reading settings
3. Running Migrations

So I put it with the others, which are technically a lie too lol, I can move the console logs into the right places, I think thats the most correct

File: migrations/20220105195635_init.ts (hunk 12 above)
Comment (mihok) [reply migrations/20220105195635_init.ts:27]: Yeah let me remove the try catch all together

File: Makefile (hunk 10 above)
Comment (mihok) [reply Makefile:55]: Anytime that we add another entry in tsconfig.js that is _outside_ of the `src/` directory it will distrupt how the `dist/` folder willl look. So if we want to revert the `dist/src/app.js` changes we would have to be okay with both the `knexfile.ts` and `migrations/` residing in `src/`, is that preferred?

Comment (evert) [reply Makefile:55]: I think so, yes!

File: src/app.ts (hunk 13 above)
Comment (evert) [reply src/app.ts:29]: Yeah can we? I imagine knex output will be interspersed with these too

File: migrations/20220105195635_init.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import { Knex } from 'knex';
+
+// Funny how this number ended up being the last changelog migration...
+const MAGIC_CHANGELOG_NUMBER = 42;
```
Comment (trasherdk) [inline migrations/20220105195635_init.ts:4]: Well.. It is the `Answer to the Ultimate Question of Life, the Universe, and Everything,` :rofl:

File: Makefile [hunk 15]
```diff
@@ -27,6 +27,15 @@ fix:
...
 lint-fix: fix
 
+migrate: build
+	npx knex migrate:latest --knexfile './dist/knexfile.ts'
+
+migrate-rollback: build
+	npx knex migrate:rollback --knexfile './dist/knexfile.ts'
+
+migration: build
+	npx knex migrate:make $(name) -x ts --knexfile './dist/knexfile.ts'
```
Comment (evert) [inline Makefile:37]:
```suggestion
```

Lets just remove this? It's less useful i think, and the name might confuse people

File: package.json [hunk 16]
```diff
@@ -7,8 +7,8 @@
     "prepublishOnly": "make build",
     "test": "make lint test",
     "tsc": "tsc",
-    "start": "node ./dist/app.js",
-    "start:watch": "tsc-watch --onSuccess 'node dist/app.js'"
+    "start": "node ./dist/src/app.js",
+    "start:watch": "tsc-watch --onSuccess 'node dist/src/app.js'"
```
Comment (evert) [inline package.json:11]: Possibly still wrong? We need to get that `src` out of there again

File: src/group/service.ts [hunk 17]
```diff
@@ -32,20 +32,17 @@ export async function addMember(group: Group, user: Principal): Promise<void> {
...
     }
-    await connection.commit();
+    await transaction.commit();
   } catch (err) {
-    await connection.rollback();
+    await transaction.rollback();
     throw err;
-  } finally {
-    connection.release();
   }
```
Comment (evert) [inline src/group/service.ts:45]: This is a red flag to me. Most database adapters will give the user a dedicated 'connection' when using `getConnection()`, until it's released back to the pool. Is this not the case with knex?

**Review by evert — CHANGES_REQUESTED** [review]
The biggest thing i can think of right now is .. do the installation instructions still make sense with the new knex migrations?

File: Makefile (hunk 15 above)
[the remark this answers is not in the document]
Comment (mihok) [reply Makefile:37]: So its okay for folks to generate their own migration files? `migrate` runs the migrations, this makes a new migration file

[the remark this answers is not in the document]
Comment (evert) [reply Makefile]: Renamed this to `create-migration`

File: src/migrations/001_initial.ts [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    table.integer('timestamp').unsigned();
+
+  });
+
+  await knex('changelog').insert({
+    id: 1,
+    timestamp: Math.floor(Date.now()/1000)
+  });
+
+  await knex.schema.createTable('users', table => {
```
Comment (evert) [inline src/migrations/001_initial.ts:22]: @mihok This is an example of a migration that I edited to work with postgres. I only did path 001 to 004 so far.

File: src/migrations/002_oauth2.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+
+  const result = await knex('changelog').select('id').where({id: 2});
+
+  if (result.length) {
+    // Migration has been applied using the old patch system
+    return;
+  }
```
Comment (evert) [inline src/migrations/002_oauth2.ts:10]: @mihok  These lines ensure that we don't re-run migrations that were run using the old system. This also means that if someone uses your 'postgres_schema/042-transfer-from-mysql.sql` file, these files also safely get skipped.

File: src/migrations/005_oauth2-redirect-uri.ts [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  if (result.length) {
+    // Migration has been applied using the old patch system
+    return;
+  }
+  await knex('changelog').insert({
+    id: 5,
+    timestamp: Math.floor(Date.now()/1000)
+  });
+
+  await knex.raw(`CREATE TABLE oauth2_redirect_uris (
```
Comment (evert) [inline src/migrations/005_oauth2-redirect-uri.ts:16]: @mihok most migrations look like this, and only work with mysql. I wrote a script to generate these files automatically.

File: .github/workflows/test.yml [hunk 21]
```diff
@@ -11,20 +11,71 @@ on:
...
     - uses: actions/checkout@v2
     - name: Use Node.js ${{ matrix.node-version }}
       uses: actions/setup-node@v1
       with:
         node-version: ${{ matrix.node-version }}
     - run: npm ci
-    - run: npm run build --if-present
-    - run: npm test
+    - run: make test
+  postres:
```
Comment (mihok) [inline .github/workflows/test.yml:52]: Improperly named I think? Given the mysql image

File: src/migrations/003_oauth-access-token.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+export async function up(knex: Knex): Promise<void> {
+
+  const result = await knex('changelog').select('id').where({id: 3});
+
+  if (result.length) {
+    // Migration has been applied using the old patch system
+    return;
+  }
+  await knex('changelog').insert({
```
Comment (mihok) [inline src/migrations/003_oauth-access-token.ts:11]: We've done away with transactions here, is this intentional?

File: src/migrations/005_oauth2-redirect-uri.ts (hunk 20 above)
Comment (mihok) [reply src/migrations/005_oauth2-redirect-uri.ts:16]: We'll have to go back through them as I think some of PostgreSQLs migration syntax is slightly different (different types) but this is a good first step having these files generated.

File: src/migrations/011_user-type.ts [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+  }
+  await knex('changelog').insert({
+    id: 11,
+    timestamp: Math.floor(Date.now()/1000)
+  });
+
+  await knex.raw(`ALTER TABLE oauth2_clients
+  DROP name`);
+  await knex.raw(`ALTER TABLE users
+  ADD type TINYINT UNSIGNED NOT NULL DEFAULT '1' COMMENT '1 = user, 2 = app'`);
```
Comment (mihok) [inline src/migrations/011_user-type.ts:19]: For example @evert this wont run in PostgreSQL because there is no `tinyint` type

**Review by mihok — COMMENTED** [review]
These changes look good @evert , its a good step to get the code in there but still rely on MySQL for now. I think once I have some freetime I'll go through the migrations and update them to work for both SQL dbs

File: .github/workflows/test.yml (hunk 21 above)
Comment (evert) [reply .github/workflows/test.yml:52]: Yes! Will fix

File: src/migrations/003_oauth-access-token.ts (hunk 22 above)
Comment (evert) [reply src/migrations/003_oauth-access-token.ts:11]: Knex wraps these migrations in something that basically works the same. The transactions in the raw sql files were helpful for certain edge cases (specifically until the fist CREATE/ALTER table that has a syntax error, but not after). I believe knex works the same.

File: src/migrations/011_user-type.ts (hunk 23 above)
Comment (evert) [reply src/migrations/011_user-type.ts:19]: Yes, but if you run the file you generated (from posgres-schema/042-migration-from-mysql.sql) then all these lines should be skipped. It's a workaround until all these migrations are rewritten but it should work

Comment (mihok) [reply src/migrations/011_user-type.ts:19]: Right good point.

**Review by evert — APPROVED** [review]
